test/helper/common.bash

```diff
@@ -1,247 +1,397 @@
+# Execute a command inside the container with name ${CONTAINER_NAME}. Moreover,
+# the command is run by Bash with `/bin/bash -c`.
+#
+# @param ...  = command to execute with Bash
+function _exec_in_container_bash() { _exec_in_container /bin/bash -c "${@}" ; }
+
+# Execute a command inside the container with name ${CONTAINER_NAME}. The command
+# is run with BATS' `run` so you can check the exit code and use `assert_`. Moreover,
+# the command is run by Bash with `/bin/bash -c`.
+#
+# @param ...  = Bash command to execute
+function _run_in_container_bash() { _run_in_container /bin/bash -c "${@}" ; }
+
+# ? << Functions to execute commands inside a container
+# ! -------------------------------------------------------------------
+# ? >> Functions about executing commands with timeouts
+
+# Repeats a given command inside a container until the timeout is over.
+#
+# @param ${1} = timeout
+# @param ${2} = container name
+# @param ...  = test command for container
+function _repeat_in_container_until_success_or_timeout() {
+  local TIMEOUT="${1:?Timeout duration must be provided}"
+  local CONTAINER_NAME="${2:?Container name must be provided}"
+  shift 2
 
-# @param ${1} target container name [IF UNSET: ${CONTAINER_NAME}]
-function get_container_ip() {
-  local TARGET_CONTAINER_NAME=${1:-${CONTAINER_NAME}}
-  docker inspect --format '{{ .NetworkSettings.IPAddress }}' "${TARGET_CONTAINER_NAME}"
+  _repeat_until_success_or_timeout \
+    --fatal-test "_container_is_running ${CONTAINER_NAME}" \
+    "${TIMEOUT}" \
+    _exec_in_container "${@}"
 }
 
-# -------------------------------------------------------------------
-
-# @param ${1} timeout
-# @param --fatal-test <command eval string> additional test whose failure aborts immediately
-# @param ... test to run
-function repeat_until_success_or_timeout {
+# Repeats a given command until the timeout is over.
+#
+# @option --fatal-test <COMMAND EVAL STRING> = additional test whose failure aborts immediately
+# @param ${1} = timeout
+# @param ...  = test to run
+function _repeat_until_success_or_timeout() {
   local FATAL_FAILURE_TEST_COMMAND
 
-  if [[ "${1}" == "--fatal-test" ]]; then
-    FATAL_FAILURE_TEST_COMMAND="${2}"
+  if [[ "${1:-}" == "--fatal-test" ]]
+  then
+    FATAL_FAILURE_TEST_COMMAND="${2:?Provided --fatal-test but no command}"
     shift 2
   fi
 
-  if ! [[ "${1}" =~ ^[0-9]+$ ]]; then
-    echo "First parameter for timeout must be an integer, received \"${1}\""
+  local TIMEOUT=${1:?Timeout duration must be provided}
+  shift 1
+
+  if ! [[ "${TIMEOUT}" =~ ^[0-9]+$ ]]
+  then
+    echo "First parameter for timeout must be an integer, received \"${TIMEOUT}\""
     return 1
   fi
 
-  local TIMEOUT=${1}
   local STARTTIME=${SECONDS}
-  shift 1
 
   until "${@}"
   do
-    if [[ -n ${FATAL_FAILURE_TEST_COMMAND} ]] && ! eval "${FATAL_FAILURE_TEST_COMMAND}"; then
+    if [[ -n ${FATAL_FAILURE_TEST_COMMAND} ]] && ! eval "${FATAL_FAILURE_TEST_COMMAND}"
```

Comment:
Whats the purpose of "eval" here, shouldn't it work flawless without too?

Reply:
I thought that too, but remember that the whole command is just one string, not an array. Using `eval` here is similar to splitting the string at whitespaces and then using the array here.